gcc/rust/resolve/rust-early-name-resolver-2.0.cc

```diff
@@ -0,0 +1,168 @@
+// Copyright (C) 2020-2023 Free Software Foundation, Inc.
+
+// This file is part of GCC.
+
+// GCC is free software; you can redistribute it and/or modify it under
+// the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3, or (at your option) any later
+// version.
+
+// GCC is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+// for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with GCC; see the file COPYING3.  If not see
+// <http://www.gnu.org/licenses/>.
+
+#include "rust-early-name-resolver-2.0.h"
+#include "rust-ast-full.h"
+#include "rust-toplevel-name-resolver-2.0.h"
+
+namespace Rust {
+namespace Resolver2_0 {
+
+Early::Early (Resolver &resolver) : DefaultResolver (resolver) {}
+
+void
+Early::go (AST::Crate &crate)
+{
+  // First we go through TopLevel resolution to get all our declared items
+  auto toplevel = TopLevel (resolver);
+  toplevel.go (crate);
+
+  rust_debug ("[ARTHUR] toplevel values:\n%s",
+	      resolver.values.as_debug_string ().c_str ());
+  rust_debug ("[ARTHUR] toplevel types:\n%s",
+	      resolver.types.as_debug_string ().c_str ());
+  rust_debug ("[ARTHUR] toplevel macros:\n%s",
+	      resolver.macros.as_debug_string ().c_str ());
+
+  textual_scope.push ();
+
+  // Then we proceed to the proper "early" name resolution: Import and macro
+  // name resolution
+  for (auto &item : crate.items)
+    item->accept_vis (*this);
+
+  textual_scope.pop ();
+}
+
+void
+Early::TextualScope::push ()
+{
+  // push a new empty scope
+  scopes.emplace_back ();
+}
+
+void
+Early::TextualScope::pop ()
+{
+  rust_assert (!scopes.empty ());
+
+  scopes.pop_back ();
+}
+
+void
+Early::TextualScope::insert (std::string name, NodeId id)
+{
+  rust_assert (!scopes.empty ());
+
+  // we can ignore the return value as we always want the latest defined macro
+  // to shadow a previous one - so if two macros have the same name and get
+  // inserted with the same key, it's not a bug
+  scopes.back ().insert ({name, id});
+}
+
+tl::optional<NodeId>
+Early::TextualScope::get (const std::string &name)
+{
+  for (auto iterator = scopes.rbegin (); iterator != scopes.rend (); iterator++)
+    {
+      auto scope = *iterator;
+      auto found = scope.find (name);
+      if (found != scope.end ())
+	return found->second;
+    }
+
+  return tl::nullopt;
+}
+
+void
+Early::visit (AST::MacroRulesDefinition &def)
+{
+  DefaultResolver::visit (def);
+
+  textual_scope.insert (def.get_rule_name ().as_string (), def.get_node_id ());
+}
+
+void
+Early::visit (AST::BlockExpr &block)
```

Comment:
Am I missing where modules are handled?

Reply:
Modules are handled directly in the `TopLevel` visitor - I don't think we have anything to do in the `Early` visitor other than resolve `use` statements to the proper node

Reply:
They need their own textual scopes, for cases like this

```rust
mod a {         // No #[macro_use]
    macro_rules! mac {
        () => {}
    }
    mac!();     // OK
}
mac!();         // ERROR - `mac` not in scope
```

Reply:
just realized that's the line you were pointing to lol. sorry I'm super tired haha